src/Nethermind/Nethermind.Merge.Plugin/InvalidChainTracker/InvalidChainTracker.cs

```diff
@@ -0,0 +1,188 @@
+    {
+        OnInvalidBlock(e.InvalidBlockHash, null);
+    }
+
+    public void SetChildParent(Keccak child, Keccak parent)
+    {
+        lock (_opLock)
+        {
+            Node parentNode = GetNode(parent);
+
+            parentNode.Children.Add(child);
+            if (parentNode.LastValidHash != null)
+            {
+                PropagateLastValidHash(parentNode);
+            }
+        }
+    }
+
+    private Node GetNode(Keccak hash)
+    {
+        if (!_tree.TryGet(hash, out Node node))
+        {
+            node = new Node();
+            _tree.Set(hash, node);
+        }
+
+        return node;
+    }
+
+    private void PropagateLastValidHash(Node node)
+    {
+        foreach (Keccak nodeChild in node.Children)
+        {
+            Node childNode = GetNode(nodeChild);
+            if (childNode.LastValidHash != node.LastValidHash)
+            {
+                childNode.LastValidHash = node.LastValidHash;
+                PropagateLastValidHash(childNode);
+            }
+        }
+    }
+
+    private BlockHeader? TryGetBlockHeader(Keccak hash)
+    {
+        if (_blockCacheService.BlockCache.TryGetValue(hash, out Block block))
+        {
+            return block.Header;
+        }
+
+        return _blockFinder.FindHeader(hash);
+    }
+
+    public void OnInvalidBlock(Keccak failedBlock, Keccak? parent)
+    {
+        if (parent == null)
+        {
+            BlockHeader? failedBlockHeader = TryGetBlockHeader(failedBlock);
+            if (failedBlockHeader == null)
+            {
+                if(_logger.IsWarn) _logger.Warn($"Unable to resolve block to determine parent. Block {failedBlock}");
+                return;
+            }
+
+            parent = failedBlockHeader.ParentHash;
+        }
+        
+        Keccak effectiveParent = parent;
+        BlockHeader? parentHeader = TryGetBlockHeader(parent);
+        if (parentHeader != null)
+        {
+            if (!_poSSwitcher.IsPostMerge(parentHeader))
+            {
+                effectiveParent = Keccak.Zero;
+            }
+        }
+        else
+        {
+            if(_logger.IsTrace) _logger.Trace($"Unable to resolve parent to determine if it is post merge. Assuming post merge. Block {parent}");
+        }
+
+        lock (_opLock)
+        {
+            Node failedBlockNode = GetNode(failedBlock);
+            failedBlockNode.LastValidHash = effectiveParent;
+            PropagateLastValidHash(failedBlockNode);
+        }
+    }
+
+    public bool IsOnKnownInvalidChain(Keccak blockHash, out Keccak? lastValidHash)
+    {
+        lock (_opLock)
+        {
+            lastValidHash = null;
+            Node node = GetNode(blockHash);
+            if (node.LastValidHash != null)
+            {
+                lastValidHash = node.LastValidHash;
+            }
+
+            return node.LastValidHash != null;
+        }
+    }
+
+    class Node
+    {
+        public HashSet<Keccak> Children { get; } = new();
```

Comment:
I think it will be easier to store Key,Value. Key - invalid block hash, Value -LastValidHash. Wdyt?

Reply:
It does that in line 178, but it also need the list of children to propagate old failure.

Reply:
I have mixed feelings about this. 

I think I would would consider doing it differently. 
Instead of keeping Children I would keep two things:
1. ParentHash
2. BlockNumber

So when propagating I would firstly sort the collection on blockNumber and then I could traverse ancestors in O(n) or O(n*ln n) level by level.

Upsides: no recursion, no potential stack issues.

Reply:
I don't understand which collection you are sorting here. The outer collection is an LRU.

Reply:
After call lets drop this its fine.